File: bader4vasp.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <regex>

// Function to read atomic positions and counts from the POSCAR file
void read_POSCAR(std::vector<std::string>& atom_types, std::vector<int>& atom_counts) {
    std::ifstream poscar_file("POSCAR");
    std::string line;

    // Read the POSCAR file line by line
    for (int i = 0; i < 7; ++i) {
        std::getline(poscar_file, line);
        if (i == 5) {
            std::istringstream iss(line);
            std::string type;
            while (iss >> type) {
                atom_types.push_back(type);
            }
        }
        else if (i == 6) {
            std::istringstream iss(line);
            int count;
            while (iss >> count) {
                atom_counts.push_back(count);
            }
        }
    }
    poscar_file.close();
}

// Function to read original charges (ZVAL) from the POTCAR file
std::map<std::string, double> read_POTCAR() {
    std::ifstream potcar_file("POTCAR");
    std::map<std::string, double> original_charges;
    std::string line;

    // Read the POTCAR file line by line
    while (std::getline(potcar_file, line)) {
        if (line.find("VRHFIN =") != std::string::npos) {
            std::regex regex_pattern("VRHFIN\\s*=\\s*(\\w+)\\s*:");
            std::smatch match;
            if (std::regex_search(line, match, regex_pattern)) {
                std::string atom_type = match[1];
                double original_charge = 0.0;

                // Look for the next lines containing 'ZVAL' for the same atom type
                while (std::getline(potcar_file, line)) {
                    if (line.find("ZVAL") != std::string::npos) {
                        std::istringstream iss(line);
                        std::string token;
                        for (int i = 0; i < 6; ++i) {
                            iss >> token; // Skip tokens until the charge value
                        }
                        try {
                            original_charge = std::stod(token);
                            original_charges[atom_type] = original_charge;
                        } catch (const std::invalid_argument& e) {
                            std::cerr << "Error converting string to double: " << token << std::endl;
                            // Handle the error, such as skipping this entry or setting a default value
                        }
                        break; // Move to the next 'VRHFIN'
                    }
                }
            }
        }
    }
    potcar_file.close();
    return original_charges;
}



// Function to read atomic charges from the ACF.dat file
std::vector<double> read_ACF_dat() {
    std::ifstream acf_file("ACF.dat");
    std::vector<double> atomic_charges;
    std::string line;

    // Read the ACF.dat file line by line
    while (std::getline(acf_file, line)) {
        if (!line.empty() && line[0] != '#') {
            std::istringstream iss(line);
            std::string field;
            for (int i = 0; i < 4; ++i) {
                iss >> field; // Skipping the first 4 fields
            }
            double charge;
            if (iss >> charge) {
                atomic_charges.push_back(charge);
            }
        }
    }
    acf_file.close();
    return atomic_charges;
}

// Function to write net charges to a file
void write_net_charges(const std::vector<std::string>& atom_types, const std::vector<int>& atom_counts,
                       const std::map<std::string, double>& original_charges, const std::vector<double>& atomic_charges) {
    std::ofstream output_file("Net_bader_charge.dat");
    output_file << "#Index Symbole Net_charge\n";

    int atom_index = 0;
    for (size_t i = 0; i < atom_types.size(); ++i) {
        std::string atom_type = atom_types[i];
        int atom_count = atom_counts[i];
        double original_charge = 0.0;

        if (original_charges.count(atom_type) > 0) {
            original_charge = original_charges.at(atom_type);
        } else {
            std::cerr << "Error: Original charge not found for atom type '" << atom_type << "'" << std::endl;
            // Handle the error, such as skipping this entry or setting a default value
        }

        for (int j = 0; j < atom_count; ++j) {
            ++atom_index;
            if (atom_index <= atomic_charges.size()) {
                double atom_charge = atomic_charges[atom_index - 1];
                double net_charge = original_charge - atom_charge;
                output_file << atom_index << " " << atom_type << " " << net_charge << "\n";
            }
            else {
                output_file << "Error: Insufficient atomic charges provided for atoms.\n";
                break;
            }
        }
    }
    output_file.close();
}

int main() {
    std::vector<std::string> atom_types;
    std::vector<int> atom_counts;
    std::map<std::string, double> original_charges;
    std::vector<double> atomic_charges;

    // Read from files
    read_POSCAR(atom_types, atom_counts);
    original_charges = read_POTCAR();
    atomic_charges = read_ACF_dat();

    // Write net charges to a file
    write_net_charges(atom_types, atom_counts, original_charges, atomic_charges);

    // Print success message
    std::cout << "Net charges calculated and saved successfully in the file 'Net_bader_charge.dat'.\n";

    return 0;
}

File: Python/bader4vasp.py
#!/usr/bin/env python3

import re
# Read atomic positions and counts from the POSCAR file
with open("POSCAR", "r") as poscar_file:
    poscar_lines = poscar_file.readlines()
    atom_types = poscar_lines[5].split()
    atom_counts = list(map(int, poscar_lines[6].split()))

# Read original charges (ZVAL) from the POTCAR file
original_charges = {}

with open("POTCAR", "r") as potcar_file:
    lines = potcar_file.readlines()

    # Iterate over each line in the POTCAR file
    i = 0
    while i < len(lines):
        line = lines[i]

        # Check if the line contains 'VRHFIN ='
        if "VRHFIN =" in line:
            # Extract the atom type using regular expression
            match = re.search(r'VRHFIN\s*=\s*(\w+)\s*:', line)
            if match:
                atom_type = match.group(1)

                # Look for the next lines containing 'ZVAL' for the same atom type
                while i < len(lines):
                    i += 1
                    next_line = lines[i]

                    # Check if the next line contains 'ZVAL'
                    if "ZVAL" in next_line:
                        # Extract the absolute charge for this atom type
                        original_charge = float(next_line.split()[5])
                        original_charges[atom_type] = original_charge
                        break  # Move to the next 'VRHFIN'

        i += 1  # Move to the next line

# Read atomic charges from the ACF.dat file
with open("ACF.dat", "r") as acf_file:
    acf_lines = acf_file.readlines()

# Skip header lines and lines starting with '#'
acf_data_lines = [line.strip() for line in acf_lines if line.strip() and not line.startswith("#")]

# Read atomic charges
atomic_charges = []
for line in acf_data_lines:
    fields = line.split()
    try:
        charge = float(fields[4])  # Extract charge from the fourth column (index 4)
        atomic_charges.append(charge)
    except (IndexError, ValueError):
        print('No chrage in this line:"', line, '". Skip it!')

# Write net charges to a file
with open("Net_bader_charge.dat", "w") as output_file:
    output_file.write("#Index Symbole Net_charge\n")
    # Assign charges to atoms based on the counts and types, and calculate net charges
    atom_index = 0
    for atom_type, atom_count in zip(atom_types, atom_counts):
        original_charge = original_charges.get(atom_type)
        if original_charge is None:
            output_file.write(f"Error: Original charge not found for atom type '{atom_type}'\n")
            continue

        for _ in range(atom_count):
            atom_index += 1
            if atom_index <= len(atomic_charges):
                atom_charge = atomic_charges[atom_index - 1]  # Match the charge with the corresponding atom
                net_charge = original_charge - atom_charge  # Subtract the atomic charge from the original charge
                output_file.write(f"{atom_index} {atom_type} {net_charge}\n")
            else:
                output_file.write("Error: Insufficient atomic charges provided for atoms.\n")
                break

# Print success message
print("Net charges calculated and saved successfully in the file 'Net_bader_charge.dat'.")

# Define a function to calculate the sum of net charges of different atoms
def calculate_net_charge(selected_atoms):
    # Convert selected atoms to a list of indices
    indices = []
    for part in selected_atoms.split(','):
        if '-' in part:
            start, end = map(int, part.split('-'))
            indices.extend(range(start, end + 1))
        else:
            indices.append(int(part))

    # Read atomic positions and counts from the POSCAR file
    with open("POSCAR", "r") as poscar_file:
        poscar_lines = poscar_file.readlines()
        atom_types = poscar_lines[5].split()
        total_atoms = sum(map(int, poscar_lines[6].split()))

    # Check if all selected indices are within the range
    max_index = total_atoms
    invalid_indices = [index for index in indices if index < 1 or index > max_index]
    if invalid_indices:
        print("Error: Selected atom indices are out of range.")
        print("Valid indices are from 1 to", max_index)
        print("Invalid indices:", invalid_indices)
        return None

    # Read Net_bader_charge.dat and calculate sum of net charges for selected atoms
    sum_charge = 0.0
    selected_atoms_list = []
    with open("Net_bader_charge.dat", "r") as charge_file:
        for line in charge_file:
            if line.startswith ("#"):
                continue
            parts = line.split()
            atom_index = int(parts[0])
            if atom_index in indices:
                sum_charge += float(parts[-1])  # Last part contains the net charge
                selected_atoms_list.append((atom_index, parts[1]))  # Atom index and type

    return sum_charge, selected_atoms_list

# Example usage:
selected_atoms = input("Enter the selected atom indices (e.g., '1,2,5' or '2-6'): ")
result = calculate_net_charge(selected_atoms)
if result is not None:
    total_charge, selected_atoms_list = result
    print("Total net charge for selected atoms:", total_charge)
    print("Atoms summed:", selected_atoms_list)


